### PR TITLE
[flang][MLIR][OpenMP] Add support for `target update` directive.

### PR DESCRIPTION
**Pull request by ergawy**
Add an op in the OMP dialect to model the `target update` direcive. This change reuses the `MapInfoOp` used by other device directive to model `map` clauses but verifies that the restrictions imposed by the `target update` directive are respected.

### CODE REVIEW

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (agozillon) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1396]: It might be worth adding a little segment explaining that the motion_operands are just a subset of the regular map flags and we still utilise the MapInfoOp to store them as in all other mapping cases currently!

**Review by agozillon — APPROVED** [review]
LGTM aside from the one Nit that can be ignored if you like. 

I'd wait for one more review before landing however!

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (agozillon) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:1403]: Ah I forgot we changed that back, thank you for the catch :D

File: mlir/test/Dialect/OpenMP/invalid.mlir
Comment (TIFitis) [inline mlir/test/Dialect/OpenMP/invalid.mlir:1667]: We don't need to add checks for cases for which it isn't possible to create an mlir op. Please remove other such checks.

Comment (ergawy) [reply mlir/test/Dialect/OpenMP/invalid.mlir:1667]: These test that we defined the op properly in TableGen, they guard against invalid op definition/change.

> ... which it isn't possible to create an mlir op.

Also, how do you guarantee that 😛?

I don't see harm in adding conservative testing.

Comment (TIFitis) [reply mlir/test/Dialect/OpenMP/invalid.mlir:1667]: The op builder wouldn't let you create an op with two nowait or if clause or whatever that isn't specified in the op definition.

By your argument why not test that it doesn't work for other random clauses as well?

Comment (ergawy) [reply mlir/test/Dialect/OpenMP/invalid.mlir]: > By your argument why not test that it doesn't work for other random clauses as well?

That I agree with. Removed the extra tests. Thanks for the discussion.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (ergawy) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:940]: I think we also need verify whether 2 separate `map_info` ops refer to the same value. I will double check that later today.

Comment (TIFitis) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:940]: When creating the updateOp, it's more likely that you'd create a new mapInfoOp along with it. However, I'm not sure if that should be an enforced rule. Can you explain why you think they should always be separate?

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (TIFitis) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:694]: I appreciate the comment but it can be a bit misleading, `implicit` isn't necessarily a map type modifier.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (TIFitis) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:936]: Did you mean to use && instead of & here?

Regardless, I think we want to check for negation of exclusive or here so it should be `if(to == from)` then emit error.

This way we can get also get rid of the other if condition below.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (clementval) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:940]: I think @TIFitis is right. You should check that your motion operands comes from MapInfoOp but I think you can have a mix of to and from. Otherwise, how to you lower smth like `!$omp target update to(a,b,c) from(x,y,z)`

Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:940]: > However, I'm not sure if that should be an enforced rule.

I am not trying to treat this as in enforced rule. What I meant is that if you have input where the **same value** is used in both `to` and `from`  `map_info` ops, we should error out. For example, see the following IR:
```
func.func @omp_target_update_invalid_motion_modifier_4(%map1 : memref<?xi32>) {
  %mapv = omp.map_info var_ptr(%map1 : memref<?xi32>, tensor<?xi32>) map_clauses(to) capture(ByRef) -> memref<?xi32> {name = ""}
  %mapv2 = omp.map_info var_ptr(%map1 : memref<?xi32>, tensor<?xi32>) map_clauses(from) capture(ByRef) -> memref<?xi32> {name = ""}

  // expected-error @below {{either to or from map types can be specified, not both}}
  omp.target_update_data motion_entries(%mapv, %mapv2 : memref<?xi32>, memref<?xi32>)
  return
}
```
According to the spec: `A list item can only appear in a to or from clause, but not in both.` ([see](https://www.openmp.org/spec-html/5.1/openmpsu69.html#x97-1070002.14.6)).

For the equivalent Fortran code:
```
subroutine foo
   integer :: a(1024)
   integer :: b(1024)
   logical :: i
   !$omp target update from(a) to(a, b) nowait
end subroutine foo
```
Here is what `flang-new` currently produces:
```
error: Semantic errors in /tmp/test.f90
/tmp/test.f90:17:29: because: 'a' appears in the FROM clause.
     !$omp target update from(a) to(a, b) nowait
                              ^
/tmp/test.f90:17:35: error: A list item ('a') can only appear in a TO or FROM clause, but not in both.
     !$omp target update from(a) to(a, b) nowait
                                    ^
/tmp/test.f90:17:35: because: 'a' appears in the TO clause.
     !$omp target update from(a) to(a, b) nowait
                                    ^
```
which is the correct behavior that we still need to reflect in the IR.

> Otherwise, how to you lower smth like `!$omp target update to(a,b,c) from(x,y,z)`

That should be totally fine. Since each `map_info` op models motion/mapping information for a **single** value, we will have one `map_info` op for each one of these variables and we are currently able to handle it fine.

Let me know if I misunderstood either of you :).

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:936]: > Did you mean to use && instead of & here?

I used `&` to be consistent with the other checks in the same function. I also don't really like it but I didn't want to divert from the surrounding code which uses `|` instead of `||`. For `bool`s both are equivalent.

> Regardless, I think we want to check for negation of exclusive or here so it should be if(to == from) then emit error.

Wouldn't it be better to have separate error messages for each situation? This is the main reason I separated them.

Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:936]: > Did you mean to use && instead of & here?

My eyes are playing tricks on me, I used `&` to be consistent with the other checks in the same function; I thought other checks in the function use the same as well. For `bool`s both are equivalent. Anyway, changed it to `&&`.

> Regardless, I think we want to check for negation of exclusive or here so it should be if(to == from) then emit error.

Wouldn't it be better to have separate error messages for each situation? This is the main reason I separated them.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp]: I updated the PR to reflect what I wrote above.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (TIFitis) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp]: These error messages would usually only be seen by working on MLIR and not end users of OpenMP, so usually we don't care about being too descriptive. However, it doesn't hurt, so you can leave them both in there if you want 😁

**Review by TIFitis — APPROVED** [review]
LGTM, Thanks for the patch!

**Pull request merged**
